tools/generate_stubs.py

```diff
@@ -0,0 +1,127 @@
+# SPDX-License-Identifier: BSD-3-Clause
```

Comment:
There are unfortunately some complications with generating stubs.
For one, our C++ package contains submodules. When `pybind11-stubgen` generates the stubs, it creates a directory structure:
```
stubs
└── scipp
   └── _scipp-stubs
      ├── __init__.pyi
      └── core
         ├── __init__.pyi
         ├── buckets
         │  └── __init__.pyi
         ├── geometry
         │  └── __init__.pyi
         └── units
            └── __init__.pyi
```
I don't know where to put those files because there is no corresponding structure in the actual source files. Hence `cpp_classes.py`.

I left out everything but `core` for now as that is the most important and lets us discuss if this is the way forward.

In addition, this script only outputs stubs for classes. Functions are wrapped in Python and don't really need extra stubs. (This means that type checking the wrappers themselves will fail. But it works fine for a user.)

And even worse: We bind a number of functions dynamically in python. They are not picked up by `pybind11-stubgen`. So I had to add them in this script. The implementation is not great. I tried using the ast directly, but before Python 3.9, we would need an extra package to turn it back into source code. So I went with manipulating strings instead.

This script is rather complicated. And it would have to be run as part of the CI to ensure that the stubs are up to date. Otherwise, we would have to remember to manually edit them every time the C++ bindings change or we bind a new function in Python.
*If* we managed to make all of scipp pass type checks, we could run a checker as part of CI and that would detect out of date stubs provided that we properly test the interface in Python. But for now, that does not seem feasible.